service/history/taskProcessingJob.go

```diff
@@ -0,0 +1,646 @@
+		return nil
+	}
+
+	// step 3: create new jobs based on result of step 2
+	newJobsOutstandingTasks := make(map[int]map[taskKey]queueTask)
+	remainingTasks := make(map[taskKey]queueTask)
+	for key, task := range t.outstandingTasks {
+		delete(t.outstandingTasks, key)
+		taskDomainID := task.Info().GetDomainID()
+		if newLevel, ok := domainIDToNewJobLevel[taskDomainID]; ok {
+			if _, ok := newJobsOutstandingTasks[newLevel]; !ok {
+				newJobsOutstandingTasks[newLevel] = make(map[taskKey]queueTask)
+			}
+			newJobsOutstandingTasks[newLevel][key] = task
+		} else {
+			remainingTasks[key] = task
+		}
+	}
+
+	var newJobs []taskProcessingJob
+	newMaxReadLevel := taskKey{
+		taskID:              t.ReadLevel().taskID + t.config.lookAheadTaskID,
+		visibilityTimestamp: t.ReadLevel().visibilityTimestamp.Add(t.config.lookAheadTimeDuration),
+	}
+	if compareTaskKeyLess(&t.maxReadLevel, &newMaxReadLevel) {
+		newMaxReadLevel = t.MaxReadLevel()
+	}
+
+	// create a job for each new level
+	newLevelToDomainIDs := make(map[int][]string)
+	for domainID, level := range domainIDToNewJobLevel {
+		newLevelToDomainIDs[level] = append(newLevelToDomainIDs[level], domainID)
+	}
+	for level, outstandingTasks := range newJobsOutstandingTasks {
+		newJob := &taskProcessingJobImpl{
+			jobLevel:         level,
+			ackLevel:         t.ackLevel,
+			readLevel:        t.readLevel,
+			maxReadLevel:     newMaxReadLevel,
+			outstandingTasks: outstandingTasks,
+			domainFilter:     newDomainFilter(newLevelToDomainIDs[level], false),
+			config:           t.config,
+			logger:           t.logger,
+			metricsClient:    t.metricsClient,
+		}
+		newJobs = append(newJobs, newJob)
+	}
+
+	// the job remaining on the current level should also be split into two parts
+	// one from ack level to read level + look ahead
+	excludedDomains := []string{}
+	for domainID := range domainIDToNewJobLevel {
+		excludedDomains = append(excludedDomains, domainID)
+	}
+	newJob := &taskProcessingJobImpl{
+		jobLevel:         t.jobLevel,
+		ackLevel:         t.ackLevel,
+		readLevel:        t.readLevel,
+		maxReadLevel:     newMaxReadLevel,
+		outstandingTasks: remainingTasks,
+		domainFilter:     t.domainFilter.exclude(excludedDomains),
+		config:           t.config,
+		logger:           t.logger,
+		metricsClient:    t.metricsClient,
+	}
+	newJobs = append(newJobs, newJob)
+
+	// the other one from read level + look ahead to max read level
+	if compareTaskKeyLess(&newMaxReadLevel, &t.maxReadLevel) {
+		newJob := newTaskProcessingJob(
+			t.jobLevel,
+			newMaxReadLevel,
+			t.maxReadLevel,
+			t.domainFilter.copy(),
+			t.config,
+			t.logger,
+			t.metricsClient,
+		)
+		newJobs = append(newJobs, newJob)
+	}
+
+	return newJobs
+}
+
+func (t *taskProcessingJobImpl) Merge(incomingJob taskProcessingJob) []taskProcessingJob {
+	incomingJobImpl, ok := incomingJob.(*taskProcessingJobImpl)
+	if !ok {
+		panic("unknown task processsing job type")
+	}
+
+	if t.jobLevel != incomingJobImpl.jobLevel {
```

Comment:
my understanding, job itself should be level / queue ID (i would prefer queue ID, since there are so many levels) independent.
each task processing queue controller, should have the queue ID.
each task processing queue, can have 0-many jobs.